### PR TITLE
upload: Make a generic interface

### PR DESCRIPTION
**Pull request by efritz**
This is a behavior-preserving refactor of the `enterprise/cmd/frontend/internal/codeintel/httpapi` package, separating the juicy, codeintel-agnostic bits into `internal/uploadhandler`. This can be instantiated by another domain, returning an `http.Handler` that does the same multipart tracking and composition into the blob store.

The API here is a little rough, but I hope we can iterate on it to polish the edges. The biggest/lowest-hanging fruit is probably a way to abstract the DBStore interface. This may not be worth it, depending on how much logic needs to be overwritten in the common cases.

## Test plan

Existing unit and end-to-end tests.

### CODE REVIEW

File: internal/uploadhandler/observability.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+type Operations struct {
+	handleEnqueue                  *observation.Operation
+	handleEnqueueSinglePayload     *observation.Operation
+	handleEnqueueMultipartSetup    *observation.Operation
+	handleEnqueueMultipartUpload   *observation.Operation
+	handleEnqueueMultipartFinalize *observation.Operation
+}
+
+func NewOperations(observationContext *observation.Context) *Operations {
```
Comment (eseliger) [inline internal/uploadhandler/observability.go:18]: should this have a name based on the handler instance? ie. uploadhandler.codeintel or something like that?

**Review by eseliger — APPROVED** [review]
awesome 🔥

File: internal/uploadhandler/observability.go (hunk 1 above)
Comment (efritz) [reply internal/uploadhandler/observability.go:18]: I actually wanted to separate it first then redo the logging and metrics completely.

Comment (eseliger) [reply internal/uploadhandler/observability.go:18]: works for me

File: enterprise/cmd/frontend/internal/codeintel/httpapi/observability.go [hunk 2]
```diff
@@ -5,39 +5,32 @@ import (
...
-	handleEnqueueMultipartFinalize *observation.Operation
+	*uploadhandler.Operations
+	authMiddleware *observation.Operation
 }
 
 func NewOperations(observationContext *observation.Context) *Operations {
 	metrics := metrics.NewREDMetrics(
 		observationContext.Registerer,
-		"codeintel_httpapi",
+		"codeintel_httpapi_auth",
```
[the remark this answers is not in the document]
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/httpapi/observability.go:19]: We actually don't graph this anywhere atm.

**Review by Gowsb — COMMENTED** [review]
H o to get html code

**Pull request merged**